runtime/include/wasm_memory.h

```diff
@@ -0,0 +1,302 @@
+#pragma once
+
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+
+#include "types.h" /* PAGE_SIZE */
+#include "wasm_types.h"
+
+#define WASM_MEMORY_MAX (size_t) UINT32_MAX + 1
+
+struct wasm_memory {
+	size_t  size;     /* Initial Size in bytes */
+	size_t  capacity; /* Size backed by actual pages */
+	size_t  max;      /* Soft cap in bytes. Defaults to 4GB */
+	uint8_t data[];
+};
+
+static inline struct wasm_memory *
+wasm_memory_allocate(size_t initial, size_t max)
```

Comment:
I've been inconsistent with the naming conventions here. I believe the follow up memory pool renames this to `wasm_memory_new`. I'm essentially following C++ semantics. "allocate" means allocate the OOP type on the heap. "init" means initialize a preallocated object, including allocating member buffers as needed. "reinit" means reinitialize members and wipe member buffers for reuse. "deinit" means cleanup and free member buffers. "free" means free the base struct from the heap. "new' means alloc + init. "delete" means deinit + free.

Reply:
I'm also generally inconsistent betwen alloc versus allocate. init versus initialize. I also sometimes name the first argument the same as the struct name and othertimes I name it self (as in Rust). Just some discussion points for things that might be work standardizing on and cleaning up in a follow-up refactor.